mmdet3d/models/necks/view_transformer.py

```diff
@@ -0,0 +1,347 @@
+
+    def create_frustum(self, depth_cfg):
+        """Generate the frustum template for each image.
+
+        Args:
+            depth_cfg (tuple(float)): Config of grid alone depth axis in format
+                of (lower_bound, upper_bound, interval).
+        """
+        H_in, W_in = self.input_size
+        H_feat, W_feat = H_in // self.downsample, W_in // self.downsample
+        d = torch.arange(*depth_cfg, dtype=torch.float)\
+            .view(-1, 1, 1).expand(-1, H_feat, W_feat)
+        self.D = d.shape[0]
+        x = torch.linspace(0, W_in - 1, W_feat,  dtype=torch.float)\
+            .view(1, 1, W_feat).expand(self.D, H_feat, W_feat)
+        y = torch.linspace(0, H_in - 1, H_feat,  dtype=torch.float)\
+            .view(1, H_feat, 1).expand(self.D, H_feat, W_feat)
+
+        # D x H x W x 3
+        frustum = torch.stack((x, y, d), -1)
+        self.frustum = nn.Parameter(frustum, requires_grad=False)
+
+    def get_lidar_coor(self, rots, trans, cam2imgs, post_rots, post_trans):
+        """Calculate the locations of the frustum points in the lidar
+        coordinate system.
+
+        Args:
+            rots (torch.Tensor): of shape (N, N_cams, 3, 3). Rotation from
+                camera coordinate system to lidar coordinate system.
+            trans (torch.Tensor): of shape (N, N_cams, 3). Translation from
+                camera coordinate system to lidar coordinate system.
+            cam2imgs (torch.Tensor): of shape (N, N_cams, 3, 3). Camera
+                intrinsic matrixes.
+            post_rots (torch.Tensor): of shape (N, N_cams, 3, 3). Rotation in
+                camera coordinate system derived from image view augmentation.
+            post_trans (torch.Tensor): of shape (N, N_cams, 3). Translation in
+                camera coordinate system derived from image view augmentation.
+
+        Returns:
+            torch.tensor: Point coordinates in shape (B, N, D, H/downsample,
+                W/downsample, 3)
+        """
+        B, N, _ = trans.shape
+
+        # post-transformation
+        # B x N x D x H x W x 3
+        points = self.frustum - post_trans.view(B, N, 1, 1, 1, 3)
+        points = torch.inverse(post_rots).view(B, N, 1, 1, 1, 3, 3)\
+            .matmul(points.unsqueeze(-1))
+
+        # cam_to_ego
+        points = torch.cat(
+            (points[..., :2, :] * points[..., 2:3, :], points[..., 2:3, :]), 5)
+        combine = rots.matmul(torch.inverse(cam2imgs))
+        points = combine.view(B, N, 1, 1, 1, 3, 3).matmul(points).squeeze(-1)
+        points += trans.view(B, N, 1, 1, 1, 3)
+
+        return points
+
+    def voxel_pooling_prepare(self, coor, x):
+        """Data preparation for voxel pooling.
+
+        Args:
+            coor (torch.tensor): of shape (B, N, D, H, W, 3). Coordinate
+                of points in the lidar space.
+            x (torch.tensor): of shape (B, N, D, H, W, C). Feature of points.
+
+        Returns:
```

Comment:
tuple[torch.Tensor]: Feature of points of shape xxx; Coordinate of points in the voxel space of shape xxx;